### PR TITLE
first steps at testing as-installed opencv

### PR DESCRIPTION
**Pull request by bollwyvl**
While looking into the built installer locally, I'm also kicking the tires on testing the OpenCV example notebook.

### CODE REVIEW

**Comment by datakurre** [comment]
Thanks. Anything that successfully imports cv2, would be enough. I have a proper internet connection for setting up virtual machine only in the Sunday evening. I’ll do final tests then.

**Comment by bollwyvl** [comment]
Looks like win could import cv2, but I should go ahead and move that test
to the conda package build rather than wait for the atest, haven't had the
chance to look at logs. No idea what to do about osx yet, might have been
flake? I have no way to test that os until at least Monday, if then. Great
timing, right? There may be some quick wins if we can get rid of qt by
depending on matplotlib-base, but I don't see how that would help the osx
issue.

**Comment by datakurre** [comment]
I can test OSX on Sunday even. The drafted release did work on OSX.

**Comment by bollwyvl** [comment]
Does open CV work on that one? Is just hard to tell, since we weren't
testing it... Perhaps I need a separate or of just the cli/nb tests, and
not the resolving.

**Comment by datakurre** [comment]
Yes, the OpenCV example worked on OSX and has been working on Windows too (I just have not tested the last build).

Did you have time to try the draft Linux installer if it worked?

My Linux test on NixOS could have been false negative and I can only get working VM on Sunday evening.

**Comment by bollwyvl** [comment]
Trying released assets inside clean vagrant boxes...

**Comment by bollwyvl** [comment]
```ruby
Vagrant.configure("2") do |config|
  config.vm.box = "ubuntu/bionic64"
  config.vm.provider "virtualbox" do |v|
    v.memory = 2048
    v.cpus = 2
  end
  config.vm.provision "shell", inline: "apt-get update"
  config.vm.provision "shell", inline: "apt-get install -y xfce4 virtualbox-guest-dkms virtualbox-guest-utils virtualbox-guest-x11"
  config.vm.provision "shell", inline: "sed -i 's/allowed_users=.*$/allowed_users=anybody/' /etc/X11/Xwrapper.config"
  config.vm.provision "file", source: "RobotLab-2020.01.0-Linux-x86_64.sh", destination: "RobotLab-2020.01.0-Linux-x86_64.sh"
  config.vm.provision "shell", inline: "bash RobotLab-2020.01.0-Linux-x86_64.sh -fbp ~/robotlab", privileged: false
end

```

**Comment by bollwyvl** [comment]
Can confirm `cv2` does not import on linux box. It looks like the name of the package changed at some point to `py-opencv`.

**Comment by bollwyvl** [comment]
It also seems not particularly possible to _install_ `py-opencv` against what's in the installer.

**Comment by datakurre** [comment]
@bollwyvl Thanks for confirming the issue. I did not bother you for nothing. Weird it still works for Windows and OSX. 

Btw, OSX test did flicker for me on Friday. Just triggered build again and it passed.

**Comment by bollwyvl** [comment]
Rad. I'll keep banging on this today.

**Comment by datakurre** [comment]
Could you check these versions? I have added in notes that 2020.01.0 has version

opencv-3.4.1-py36h6fd60c2_1.tar.bz2

But 2019.12.0 has version

opencv-3.4.2-py36h6fd60c2_1.tar.bz2

That’s curious.

**Comment by datakurre** [comment]
Anyway, maybe this could be fixed by just finding opencv version that includes Python bindings for all platforms. Linux missing cv2 may be just packaging error 🤔

**Comment by bollwyvl** [comment]
Yeah, I think adding matplotlib did some _things_ to the final resolution.

**Comment by bollwyvl** [comment]
Ick: osx couldn't find the right FF during the build...

**Comment by bollwyvl** [comment]
Ok, all the OSX failures are now due to `headlessfirefox`... in the robotlab tests, we're having to do
```robotframework
Create WebDriver    Firefox    executable_path=${geckodriver}    firefox_binary=${firefox}
```

I'm thinking that patching the notebook on packaging will be the most robust solution, but it's going to take a sec.

The windows fail looks like simple flake.

File: recipes/robotlab/builder.py [hunk 1]
```diff
@@ -19,6 +21,13 @@
 }
 SVG = SRC_DIR / "robotlab" / "src" / "robotlab" / "icons" / "starter.svg"
 
+OPENCV_OSX = """
```
Comment (bollwyvl) [inline recipes/robotlab/builder.py:24]: Not prod of this: any thoughts of ways to make it cleaner? Can we add something at the RFSL layer?

File: recipes/robotlab/builder.py [hunk 2]
```diff
@@ -33,6 +42,20 @@
...
+    opencv = EXAMPLES / "OpenCV.ipynb"
+    nb = json.loads(opencv.read_text())
+    for cell in nb["cells"]:
+        if cell["cell_type"] == "code":
+            lines = cell["source"]
+            for i, line in enumerate(lines):
+                if line.startswith("Open Browser"):
+                    lines[i] = OPENCV_OSX
+
+    opencv.write_text(json.dumps(nb, indent=2))
```
Comment (bollwyvl) [inline recipes/robotlab/builder.py:59]: yeesh

**Comment by datakurre** [comment]
I am ok with search and replacing headlessfirefox with firefox if that helps for now.

**Comment by bollwyvl** [comment]
nah, there's hardcoded stuff to look for it in like
Applications/FireFox.app/Contents/Blah/Blah i think... thing is it actually
started it, but then crapped out in some opencv stuff.

**Comment by bollwyvl** [comment]
C'mon windows, no :snowflake:

**Comment by bollwyvl** [comment]
:laughing:

**Comment by datakurre** [comment]
No sleep? 🙃

I’ll try these out in 12 hours.

**Comment by datakurre** [comment]
*  [x] MacOS
*  [x] Linux
*  [x] Windows

**Comment by datakurre** [comment]
What is the correct way to activate the env in Linux and MacOS.

```
source bin/activate
```
?

**Comment by bollwyvl** [comment]
Yeah that's what we do generally...

**Comment by datakurre** [comment]
That was quite a patch you had to do for OSX on OpenCV demo, and those patched notebooks ended up in the final distribution. It's ok for now, but need to think about those. I'd feel better to have simpler test notebooks separately for selenium and OpenCV in RobotLab instead of trying to execute everything crazy I dump into robotkernel examples :-)

**Comment by datakurre** [comment]
I learned the way to test Linux-installer on top of NixOS. OpenCV works.

Once again you saved the day!

**Comment by datakurre** [comment]
All green. Going to merge and relabel "More on robotkernel"-starter. And then move v2020.01.0 tag.

**Comment by bollwyvl** [comment]
Agreed, too crazy, but presumably would be required to work for a robotlab
user on osx that didn't have ff already. But the more we can test, the less
we have to manually test... Or in my case with osx, the only way I care to
test. It's one of the key stories I tell people about this project.

I could see offering a Library that did some of these dirty tricks. But wanted
to get something working first.

***Settings***
Library SeleniumLibrary
Library robotlab.FirefoxAsHeadlessFirefox

Open browser http://Google.com browser=headlessfirefox

And just do a savage hack of selenium to do our bidding. Wouldn't want it
to be automatic, though.

**Comment by bollwyvl** [comment]
Hooray!

**Comment by datakurre** [comment]
Needless to say I do appreciate a lot your effort and all the tests! There is your name and picture on the slides :-) https://github.com/robots-from-jupyter/robotkernel/blob/master/docs/robocon2020/index.ipynb This time I've prepared them with Notebook + RISE.

**Comment by datakurre** [comment]
Tested and released. Thanks again! (Windows tests did flicker once.)

**Comment by bollwyvl** [comment]
With _hours_ to spare!

**Comment by datakurre** [comment]
Hours... and a day! The workshop is only in 21 hours.

**Comment by bollwyvl** [comment]
Heck, then what was I even worried about! But it's still great we got it
out there.

**Comment by datakurre** [comment]
It is. Now I can have a good night sleep before (or try out what's broken in the latest RF 3.2 beta). And I did share the download links with the participants already in the morning.

**Pull request merged**